### PR TITLE
feat: native light client state verification

### PR DESCRIPTION
**Pull request by xavierdmello**
Inside the Vector pallet, instead of verifying a proof of light client logic, we natively implement the logic at a node level. This mirrors [SP1 Telepathy's](https://github.com/succinctlabs/sp1-telepathy/) approach. The Rust-based logic built on top of [Helios](https://github.com/a16z/helios) allows for seamless integration with Avail's pallet.

Additionally, this approach dramatically simplifies the operator. Instead of fetching specialized inputs and generating proofs, we simply retrieve inputs for Helios and relay that to the chain.

## Modifications
We've optimized storage by directly using the canonical Ethereum consensus state's sync committee hash, eliminating the need for Poseidon hashes.

The updated verification logic is inside the `fulfill_call` function of  [`pallets/vector/src/lib.rs`](https://github.com/xavierdmello/avail/blob/xavier/helios-light-client/pallets/vector/src/lib.rs).

## Operator
The new operator is based on [SP1 Telepathy's operator](https://github.com/succinctlabs/sp1-telepathy/blob/dc08bae2e08cfdc8d795c34c6fbd5d7e5cf917bd/script/bin/operator.rs). The logic remains the same, with placeholders for reading and writing to Avail using `avail-subxt` instead of a smart contract on Ethereum.

For reference, I've created a fork of the SP1 Telepathy operator with added comments highlighting the lines that need to be changed: https://github.com/xavierdmello/avail-telepathy-operator

## Testing
We've adapted a few tests into `tests_new.rs`, which replaces the current `tests.rs`. You can try the new ones by running `cargo test test_fulfill` inside of `./pallets/vector/`. We generated test inputs using SP1 Telepathy ([instructions here](https://github.com/succinctlabs/sp1-telepathy#generating-hardcoded-test-cases)), which outputs bytes that should be placed inside `./pallets/vector/examples`

### CODE REVIEW

File: pallets/vector/src/lib.rs [hunk 1]
```diff
@@ -395,72 +410,188 @@ pub mod pallet {
...
+					&& verify_update(
+					update,
+					expected_current_slot,
+					&store,
+					genesis_root.clone(),
+					&forks,
+				)
+					.is_ok();
+
+				apply_update(&mut store, update);
```
Comment (fmiguelgarcia) [inline pallets/vector/src/lib.rs:478]: That will apply the `update` even if it was invalid. 
Do we want this behavior? Should not we raise up an error in that case (short-circuit)?

File: pallets/vector/src/lib.rs [hunk 2]
```diff
@@ -395,72 +410,188 @@ pub mod pallet {
...
+			is_valid = is_valid
+				&& verify_finality_update(
+				&finality_update,
+				expected_current_slot,
+				&store,
+				genesis_root.clone(),
+				&forks,
+			)
+				.is_ok();
+			apply_finality_update(&mut store, &finality_update);
```
Comment (fmiguelgarcia) [inline pallets/vector/src/lib.rs:491]: Same questions here about applying invalid update.

File: pallets/vector/src/lib.rs [hunk 3]
```diff
@@ -395,72 +410,188 @@ pub mod pallet {
...
+
+			// 3. Verify execution state root proof
+			let execution_state_branch_nodes: Vec<Node> = execution_state_proof
+				.execution_state_branch
+				.iter()
+				.map(|b| Node::try_from(b.as_ref()).unwrap())
```
Comment (fmiguelgarcia) [inline pallets/vector/src/lib.rs:497]: ditto about `unwrap` to `Error`

File: pallets/vector/src/lib.rs [hunk 4]
```diff
@@ -395,72 +410,188 @@ pub mod pallet {
...
+				.collect();
+
+			is_valid = is_valid
+				&& is_valid_merkle_branch(
+				&Node::try_from(execution_state_proof.execution_state_root.as_ref()).unwrap(),
```
Comment (fmiguelgarcia) [inline pallets/vector/src/lib.rs:502]: ditto about `unwrap` to `Error`.

File: pallets/vector/src/lib.rs [hunk 5]
```diff
@@ -395,72 +410,188 @@ pub mod pallet {
...
+				execution_state_branch_nodes.iter(),
+				MERKLE_BRANCH_DEPTH,
+				MERKLE_BRANCH_INDEX,
+				&Node::try_from(store.finalized_header.body_root.as_ref()).unwrap(),
```
Comment (fmiguelgarcia) [inline pallets/vector/src/lib.rs:506]: ditto about `unwrap` to `Error`.

File: pallets/vector/src/lib.rs [hunk 6]
```diff
@@ -395,72 +410,188 @@ pub mod pallet {
...
+			);
+
+			let finalized_header_root: [u8; 32] = store
+				.finalized_header
+				.hash_tree_root()
+				.unwrap().as_ref().try_into().unwrap();
```
Comment (fmiguelgarcia) [inline pallets/vector/src/lib.rs:512]: ditto about `unwrap`.

File: pallets/vector/src/lib.rs [hunk 7]
```diff
@@ -395,72 +410,188 @@ pub mod pallet {
...
 
-			// make sure that verification call is valid
-			ensure!(is_success, Error::<T>::VerificationFailed);
+			// 4. Store step if needed
+			if prev_head != head {
+				let verified_output = VerifiedStepOutput {
+					finalized_header_root: H256::from(finalized_header_root),
+					execution_state_root: H256::from(execution_state_root),
+					finalized_slot: store.finalized_header.slot.as_u64(),
+					participation: store.current_max_active_participants.try_into().unwrap(),
```
Comment (fmiguelgarcia) [inline pallets/vector/src/lib.rs:531]: ditto about `unwrap`

File: pallets/vector/src/lib.rs [hunk 8]
```diff
@@ -395,72 +410,188 @@ pub mod pallet {
...
+			// a) Store current sync committee if stored one is empty (i.e. first time or after a range of updates)
+			let period = head.as_u64()
+				.checked_div(config.slots_per_period)
+				.ok_or(Error::<T>::ConfigurationNotSet)?;
+			let stored_current_sync_committee = SyncCommitteeHashes::<T>::get(period);
+			if stored_current_sync_committee.is_zero() {
+				let current_sync_committee_hash: U256 = store
+					.current_sync_committee
+					.hash_tree_root()
+					.unwrap()
```
Comment (fmiguelgarcia) [inline pallets/vector/src/lib.rs:557]: ditto about `unwrap`.

File: pallets/vector/src/lib.rs [hunk 9]
```diff
@@ -395,72 +410,188 @@ pub mod pallet {
...
+				function_called = true;
+
+				Self::set_sync_committee_hash(period, current_sync_committee_hash)?;
+			}
+
+			// b) Store next sync committee if available
+			if let Some(mut next_sync_committee) = store.next_sync_committee {
+				let next_period = period + 1;
+				let stored_next_sync_committee_hash = SyncCommitteeHashes::<T>::get(next_period);
+				let next_sync_committee_hash: [u8; 32] = next_sync_committee
```
Comment (fmiguelgarcia) [inline pallets/vector/src/lib.rs:574]: ditto about `unwrap`.

File: pallets/vector/src/lib.rs [hunk 10]
```diff
@@ -368,7 +383,7 @@ pub mod pallet {
 					.expect("Rotate verification key should be valid at genesis.");
 			RotateVerificationKey::<T>::set(Some(rotate_verification_key));
 
-			SyncCommitteePoseidons::<T>::insert(self.period, self.sync_committee_poseidon);
+			SyncCommitteeHashes::<T>::insert(self.period, self.sync_committee_hash);
```
Comment (fmiguelgarcia) [inline pallets/vector/src/lib.rs:386]: The mainnet genesis will be different using this new binary even if we use the same chain spec. 
I am not sure about side-effect of that.

**Review by fmiguelgarcia — COMMENTED** [review]
Some comments and questions.

**Comment by ToufeeqP** [comment]
closing in favour of #689

**Pull request closed without merging**